avldb/src/main/scala/scorex/crypto/authds/avltree/batch/ProverNodeSerializer.scala

```diff
@@ -0,0 +1,71 @@
+package scorex.crypto.authds.avltree.batch
+
+import com.google.common.primitives.Ints
+import scorex.core.serialization.ErgoSerializer
+import scorex.crypto.authds.{ADKey, ADValue, Balance}
+import scorex.crypto.authds.avltree.batch.Constants.{DigestType, hashFn}
+import scorex.crypto.authds.avltree.batch.serialization.ProxyInternalNode
+import scorex.crypto.hash.Digest32
+import scorex.db.LDBVersionedStore
+import scorex.util.serialization.{Reader, Writer}
+
+/**
+  * ErgoSerializer based prover nodes serializer (alternative to one provided in scrypto which is doing much more
+  * allocations)
+  *
+  * @param store - store which could be provided to fetch children of a node when a child is requested
+  */
+class ProverNodeSerializer(store: LDBVersionedStore) extends ErgoSerializer[ProverNodes[DigestType]] {
+
+  import VersionedLDBAVLStorage.{InternalNodePrefix,LeafPrefix}
+
+  override def serialize(node: ProverNodes[DigestType], w: Writer): Unit = {
+    node match {
+      case n: ProxyInternalNode[DigestType] =>
+        w.put(InternalNodePrefix)
+        w.put(n.balance)
+        w.putBytes(n.key)
+        w.putBytes(n.leftLabel)
+        w.putBytes(n.rightLabel)
+      case n: InternalProverNode[DigestType] =>
+        w.put(InternalNodePrefix)
+        w.put(n.balance)
+        w.putBytes(n.key)
+        w.putBytes(n.left.label)
+        w.putBytes(n.right.label)
+      case n: ProverLeaf[DigestType] =>
+        w.put(LeafPrefix)
+        w.putBytes(n.key)
+        w.putBytes(Ints.toByteArray(n.value.length))
+        w.putBytes(n.value)
+        w.putBytes(n.nextLeafKey)
```

Comment:
You are proposing consensus-breaking change, serialization has to be fixed to be compatible with previous versions, otherwise there will be disagreement on block header even (`stateRoot` field)

Reply:
Ops, missed this, good point.